python/cudf_polars/cudf_polars/dsl/translate.py

```diff
@@ -302,13 +302,37 @@ def _(
     # Join key dtypes are dependent on the schema of the left and
     # right inputs, so these must be translated with the relevant
     # input active.
+    def adjust_literal_dtype(literal: expr.Literal) -> expr.Literal:
+        if literal.dtype.id() == plc.types.TypeId.INT32:
+            plc_int64 = plc.types.DataType(plc.types.TypeId.INT64)
+            return expr.Literal(
+                plc_int64,
+                pa.scalar(literal.value.as_py(), type=plc.interop.to_arrow(plc_int64)),
+            )
+        return literal
+
+    def maybe_adjust_binop(e) -> expr.Expr:
+        if isinstance(e.value, expr.BinOp):
+            left, right = e.value.children
+            if isinstance(left, expr.Col) and isinstance(right, expr.Literal):
+                e.value.children = (left, adjust_literal_dtype(right))
+            elif isinstance(left, expr.Literal) and isinstance(right, expr.Col):
+                e.value.children = (adjust_literal_dtype(left), right)
+        return e
+
+    def translate_expr_and_maybe_fix_binop_args(translator, exprs):
```

Comment:
I'm doing this change because integer literals default to int32 I think.

Reply:
I'll defer to @wence- here.

Reply:
Again, I think this should not be necessary at all because the polars IR should be typed by now. Can you show an example of what breaks?

Reply:
Hmm, I think this is a bug in the polars type-coercion phase. For a normal (not `join_where`) join, it inserts the cast.

Reply:
I opened https://github.com/pola-rs/polars/issues/20935